Add map_index support to all task instance-related views

uranusjr opened the pull request:
This makes multiple ti-related views take an additional parameter `map_index` (optional, defaults to -1 if not given for unmapped tasks). But I’m not really sure how to actually _pass_ the argument in those. Also `map_index` is added to `list_columns` for views to models with that field.

Some oddities not handled in this PR that need to be fixed later:

* `/success` and `/failed` set state to a task (and potentially its upstreams/downstreams). It feels a bit weird to be able to set the state to only one mapped task? Also upstream/downstream makes less sense if any of those are mapped. So I’m currently leaving these views alone (i.e. setting success/failed to a mapped task sets the state of all mapped tis). We can perhaps add some UI components to optionally do more fine-grained state modification.
* `SlamissModelView` is currently unchanged awaiting #22184.
* `LogModelView` is currently unchanged awaiting #22274.

uranusjr committed:
Add map_index support to rendered-templates

uranusjr committed:
Add map_index support to rendered-k8s

uranusjr committed:
Add map_index support to get_logs_with_metadata

uranusjr committed:
Add map_index support to log

uranusjr committed:
Add map_index support to redirect_to_external_log

uranusjr committed:
Add map_index support to task

uranusjr committed:
fixup! Add map_index support to task

uranusjr committed:
Add map_index support to xcom

uranusjr committed:
Add map_index support to run

uranusjr committed:
Add map_index support to extra_links

uranusjr committed:
Add map_index to model views

uranusjr committed:
Pass map_index to url_for() in templates

This does not account for js-generated URLs.

uranusjr committed:
Pass map index to JavaScript

uranusjr committed:
fixup! Add map_index to model views

bbovenzi commented on airflow/www/static/js/dag.js:
```diff
@@ -53,6 +53,7 @@ let taskId = '';
 let executionDate = '';
 let subdagId = '';
 let dagRunId = '';
+let mapIndex = undefined;
```
```suggestion
let mapIndex = -1;
```
```suggestion
let mapIndex;
```

The linter doesn't like initializing as `undefined` so let's default to `-1` or not set anything

ashb replied:
I'm on this.

bbovenzi commented on airflow/www/static/js/gantt.js:
```diff
@@ -204,7 +204,7 @@ d3.gantt = () => {
       .on('mouseout', tip.hide)
       .on('click', (d) => {
-        callModal(d.task_id, d.execution_date, d.extraLinks, undefined, undefined, d.run_id);
+        callModal(d.task_id, d.execution_date, d.extraLinks, undefined, undefined, d.run_id, d.map_index);
```
```suggestion
        // eslint-disable-next-line max-len
        callModal(d.task_id, d.execution_date, d.extraLinks, undefined, undefined, d.run_id, d.map_index);
```

linter is angry about this. Let's ignore it for now. In a later PR, I would like to use object props here to avoid passing `undefined` and make it clearer which arg is what.

bbovenzi commented on airflow/www/static/js/graph.js:
```diff
@@ -172,8 +172,8 @@ function draw() {
-      if (task.task_type === 'SubDagOperator') callModal(nodeId, executionDate, task.extra_links, tryNumber, true, dagRunId);
-      else callModal(nodeId, executionDate, task.extra_links, tryNumber, undefined, dagRunId);
+      if (task.task_type === 'SubDagOperator') callModal(nodeId, executionDate, task.extra_links, tryNumber, true, dagRunId, task.map_index);
+      else callModal(nodeId, executionDate, task.extra_links, tryNumber, undefined, dagRunId, task.map_index);
```
We probably need the `// eslint-disable-next-line max-len` or to split everything into multiple lines.

ashb committed:
Don't put map_index in URLS (from JS) for unmapped tasks

`map_index=-1` is an implementation detail of "this is not a mapped
task" that we shouldn't really expose to UI users.

ashb committed:
Don't include `map_index=-1` in Jinja generated URLs either

Sadly we can't to `url_for(route, **params)` in Jinja, so we have to
"duplicate" it like this

ashb commented on airflow/www/static/js/dag.js:
```diff
@@ -151,46 +163,41 @@ export function callModal(t, d, extraLinks, tryNumbers, sd, drID) {
-    }&task_id=${encodeURIComponent(taskId)
-    }&execution_date=${encodeURIComponent(executionDate)
-    }&metadata=null`
-      + '&format=file';
```
Whoops, lost this. One moment

ashb committed:
fixup! Don't put map_index in URLS (from JS) for unmapped tasks

ashb committed:
fixup! Don't put map_index in URLS (from JS) for unmapped tasks

uranusjr commented on airflow/www/static/js/graph.js:
```diff
@@ -172,8 +172,15 @@ function draw() {
+        executionDate,
+        task.extra_links,
+        tryNumber,
+        task.task_tupe === 'SubDagOperator',
```
```suggestion
        task.task_tupe === 'SubDagOperator' ? true : undefined,
```

Does it make a difference? This is closer to the original.

ashb replied:
Yes it is, but it's not needed now because i changed how this parameter is handled

uranusjr commented on airflow/www/static/js/tree/StatusBox.jsx:
```diff
@@ -36,9 +36,9 @@ const StatusBox = ({
+    executionDate, taskId, tryNumber = 0, operator, runId, mapIndex,
   } = instance;
-  const onClick = () => executionDate && callModal(taskId, executionDate, extraLinks, tryNumber, operator === 'SubDagOperator' || undefined, runId);
+  const onClick = () => executionDate && callModal(taskId, executionDate, extraLinks, tryNumber, operator === 'SubDagOperator', runId, mapIndex);
```
Same, does it matter to use false vs undefined?

ashb commented on airflow/www/static/js/dag.js:
```diff
@@ -115,26 +124,29 @@ document.addEventListener('click', (event) => {
     subdagId = `${dagId}.${t}`;
+  } else {
+    $('#div_btn_subdag').hide();
+    subdagId = undefined;
```
@uranusjr see this change for your two comments

ashb committed:
Show map index in TI details and modal

bbovenzi commented on airflow/www/static/js/dag.js:
```diff
@@ -135,12 +135,18 @@ export function callModal(t, d, extraLinks, tryNumbers, sd, drID, mi) {
   $('#extra_links').empty().hide();
+  if (mi >= 0) {
+    $('#modal_map_index').show();
+    $('#map_index, #modal_map_index .value').text(mi);
```
Where is `#map_index`? I don't see that id anywhere

ashb replied:
Hmmm good question. This was already there, I just moved it. This might be setting the form control.

uranusjr replied:
I think `#` also matches `name` attributes. (Don’t take my words, I’m not fluent in js by any means.)

bbovenzi replied:
Oh yes, that is how everything else is working. My bad.

ashb replied:
It wasn't needed though, so I've removed this new addition, (left the second selector here)

bbovenzi commented on airflow/www/templates/airflow/task_instance.html:
```diff
@@ -27,23 +27,33 @@
   <h4>
     <span class="text-muted">Task Instance:</span> <span>{{ task_id }}</span>
     <span class="text-muted">at</span> <time datetime="{{ execution_date }}">{{ execution_date }}</time>
+    {% if map_index and map_index != -1 %}
```
I think we need to make sure that `map_index` is updating.
In Gantt:
- view mapped instance -> click details -> see map Index
- view a normal instance's details first - > repeat the flow above -> no map index

ashb replied:
I've been testing this by hacking `ti.map_index=5` in the `def task` in the view (the task instance details view)

ashb replied:
Not quite working right though.

ashb committed:
Correctly submit map_index in Task modal forms

ashb commented on airflow/www/templates/airflow/task_instance.html:
```diff
@@ -27,23 +27,33 @@
   <h4>
     <span class="text-muted">Task Instance:</span> <span>{{ task_id }}</span>
     <span class="text-muted">at</span> <time datetime="{{ execution_date }}">{{ execution_date }}</time>
+    {% if map_index and map_index != -1 %}
```
```suggestion
    {% if map_index|defined and map_index != -1 %}
```

I think

bbovenzi replied:
Fixed with `is defined`

bbovenzi committed:
Fix jinja if statement

uranusjr commented on airflow/www/templates/airflow/task_instance.html:
```diff
@@ -27,23 +27,33 @@
   <h4>
     <span class="text-muted">Task Instance:</span> <span>{{ task_id }}</span>
     <span class="text-muted">at</span> <time datetime="{{ execution_date }}">{{ execution_date }}</time>
+    {% if map_index is defined and map_index != -1 %}
```
```suggestion
    {% if map_index is defined and map_index < 0 %}
```

uranusjr reviewed (commented):
Instead of checking for -1 specifically, check for all negative map indexes. This is more in line with implementation elsewhere.

uranusjr committed:
Check for negative map index instead of -1 specifically

This is more in line with implementation elsewhere.
